### ISSUE TITLE
GRANDPA keys change every session, but BEEFY keys shouldn't

### ISSUE THREAD

**Issue by tomusdrw** [paritytech/grandpa-bridge-gadget#9]
Currently session is 4h on polkadot, and that's the cadence of GRANDPA keys changing.
The validator set however only changes every era (epoch), which is 4 sessions (around 16h).

We should make sure that BEEFY keys stay the same across the entire epoch (i.e. to have one set transition block per epoch, not per era).

CC @andresilva

**Comment by seunlanlege** [comment]
dumb question, but wouldn't this mean that GRANDPA keys will diverge from BEEFY keys?

**Comment by tomusdrw** [comment]
Session keys are already different for GRANDPA and BEEFY (it's even different crypto), but the validator set stays the same (i.e. the slashable actors). So this change is not about diverging on the validator set members level, but just the session keys they are using - i.e. for BEEFY they wouldn't change the session keys that frequently.

**Comment by seunlanlege** [comment]
Right,wouldn't that bring up extra security concerns if the BEEFY session key never changes in between sessions?

Is this because there's a risk BEEFY might not complete a round in 4h?

**Comment by tomusdrw** [comment]
> Right,wouldn't that bring up extra security concerns if the BEEFY session key never changes in between sessions?

This does not influence the security of the protocol (same bonded actors) - the point of changing session keys frequently is to reduce the risks in case they get compromised (i.e. validator leaks their keys and someone can use them to vote on alternative forks and/or trigger equivocations). So it makes it a bit less convenient (or more risky) for the validators.

> Is this because there's a risk BEEFY might not complete a round in 4h?

Not really, BEEFY should be completing rounds as quickly as possible assuming validators are online and network conditions are good, ideally every 2-3 blocks, similarly as GRANDPA.
The point is to reduce the minimal frequency of light client updates. Now it's once per session, because the session keys change - i.e. a light client has to import at least one BEEFY justification every session. With this reduction, a light client could only import stuff once per era which may significantly reduce the costs (obviously increasing latency though).

**Comment by seunlanlege** [comment]
> So it makes it a bit less convenient (or more risky) for the validators.

Yeah this is what I meant.

> With this reduction, a light client could only import stuff once per era which may significantly reduce the costs (obviously increasing latency though).

Personally I think mandating that light clients import a new header every session (4h) is fine and 4h is already a high enough latency, 16h just makes the bridge unusable.

**Comment by andresilva** [comment]
> Yeah this is what I meant.

Keys need to be rotated by humans, it's not something we can force them to do. And I'm pretty sure most validators aren't actively rotating their session keys every 4 hours.

> Personally I think mandating that light clients import a new header every session (4h) is fine and 4h is already a high enough latency, 16h just makes the bridge unusable.

The point is only having to bridge at least one block per era, it doesn't mean it can't be done more often. And the bridge will be trustless, anyone that doesn't want to wait on the relayers cadence can bridge a finality proof themselves whenever they want (and pay for it).

**Comment by seunlanlege** [comment]
> The point is only having to bridge at least one block per era, it doesn't mean it can't be done more often

Yeah I may have misunderstood, I thought BEEFY may only produce blocks once per era. But yeah I guess this is just to drop the requirement on bridge implementations. It's a nice to have.

**Comment by seunlanlege** [comment]
> We should make sure that BEEFY keys stay the same across the entire epoch (i.e. to have one set transition block per epoch, not per era).

How do we make this happen? Implementation details.

cc @andresilva @tomusdrw

**Comment by seunlanlege** [comment]
👀

**Comment by acatangiu** [comment]
Stale issue. We've decided we want to have one BEEFY mandatory block per `pallet_session` session, and BEEFY validator set changes each session.
Validator set id is incremented each session, while validator keys only change every era (epoch).
